### ISSUE TITLE
Lots of whitespace differences introduced by endo bundler

### ISSUE THREAD

**Issue by erights** [endojs/endo#926]
On the right is the original `lockdown-shim.js`. On the left is the corresponding portion of `lockdown.umd.js` aligned for a visual line by line comparison. There are lots of little divergences. I do not remember seeing these divergences when we were using some previous bundler (rollup?).

These differences will impede the debugging experience. It means that at least the columns are different, and so we cannot simply sourceURL the original source file to appear in the debugger in lieu of the bundled source.

![image](https://user-images.githubusercontent.com/273868/140021609-2fa99c34-99e3-45c4-8268-ee25e0e67d44.png)

Related issues:
* https://github.com/endojs/endo/issues/400

**Comment by kriskowal** [comment]
We discussed this and resolved that this probably can be addressed by patching the babel-generator code that catches up with the cursor of the next written AST, such that it advances the column in addition to the line.

**Comment by kriskowal** [comment]
We also discussed trying to preserve untouched subtrees of the AST by reusing the original source. This is possible but more invasive. It requires annotating which nodes were “touched” and modifying bundle-source to attend to that detail.

**Comment by pyramation** [comment]
I would encourage the use of source maps vs. expecting babel to allow customization of whitespace:

https://github.com/evanw/node-source-map-support
https://github.com/chocolateboy/babel-plugin-source-map-support

If you integrate source maps correctly, all debugging should theoretically use the source code files instead of the transpiled code in stack traces.

If possible, I'd encourage and suggest finding a path that is not dependent on comments or whitespace output of babel. As far as I understand the deparser (babel/generate) or the transformer, whitespace (and comments) are two areas that I've empirically found that babel doesn't address perfectly.

Given the dependency on babel (which is the industry standard), if sourcemaps don't solve this issue, I'd suggest making an issue in babel. Looks like it has been brought up before as early as 2015, but has been closed https://github.com/babel/babel/issues/497

**Comment by pyramation** [comment]
also, looking more deeply into the original babel issue I found a linked issue that helped me find this library: https://github.com/js-cst-tokens/cst-tokens 

they are not using ASTs, but rather, CSTs which appear to preserve whitespace. I would still think sourcemaps are the better path, but figured I'd share this link.

**Comment by erights** [comment]
Thanks! If the cst works (it looks like a great find!), I prefer it to the sourcemap path. With sourcemaps, the programmer is often looking at their original source text which can differ from the code actually being executed in devious ways. Even with the cst, I'd expect programmers to often be looking at their original source test, via sourceURL rather than source text. But, since these are supposed to differ only on `import` and `export`, 
   * programmers can avoid the sourceURL view and have essentially the same experience
   * differences on lines other than `import` and `export` can be trivially caught by a diff.

CST FTW! Thanks much. I hope this works!

All that said, if we do sourcemaps, we can easily write a checker to ensure that the pre vs post source text differs only on `import`, `export`, and whitespace. We should probably do that anyway. Once that's verified, debugging through a sourcemap view would be safe.

**Comment by erights** [comment]
Actually, the current babel output seems to differ on more than whitespace, for example unnecessary parens, so that a correspondence verifier for babel output is no longer such a trivial exercise. This pushes harder for us to convert to CSTs.

**Comment by pyramation** [comment]
makes sense! From what I can tell, the intention of CSTs is more aligned with your goals.

From what I can see on the readme, the flow should be somewhat similar to babel. So if you guys have a babel transformer for `import` and `export`, it should be decently portable to take a crack at using CSTs via these newly discovered libraries.

Keep me in the loop!

**Comment by erights** [comment]
> makes sense! From what I can tell, the intention of CSTs is more aligned with your goals.
> 
> From what I can see on the readme, the flow should be somewhat similar to babel. So if you guys have a babel transformer for `import` and `export`,

We do!

> it should be decently portable to take a crack at using CSTs via these newly discovered libraries.
> 
> Keep me in the loop!

Will do. Much appreciated.

**Comment by erights** [comment]
See also https://github.com/endojs/endo/issues/1044

**Comment by kriskowal** [comment]
I’ve threaded source maps into a per-user cache as of #1689. This is as much as we can do to improve debugging a module through the module-to-program transform in our shim until we hear back from @nicolo-ribaudo.

**Comment by erights** [comment]
Another one. In order to send a breakpoint in the `handle` method shown in my source on the left panel, I waited until the debugger took me into the corresponding bundled file shown on the right, line them up (which is why there's a very long first line scrolled off the top on the right). Fortunately, the line numbers do correspond accurately, so with these lined up I can interact with the debugger on the right pane while looking at the left pane.

![image](https://github.com/endojs/endo/assets/273868/64b09a56-96e4-4017-be8a-c178c4a85d24)
